File: ios/Classes/RNKLineView.m
#import "RNKLineView-Swift.h"
#import "RCTViewManager.h"


@interface RCT_EXTERN_MODULE(RNKLineView, RCTViewManager)

RCT_EXPORT_VIEW_PROPERTY(onDrawItemDidTouch, RCTBubblingEventBlock)

RCT_EXPORT_VIEW_PROPERTY(onScrollLeft, RCTBubblingEventBlock)

RCT_EXPORT_VIEW_PROPERTY(onChartTouch, RCTBubblingEventBlock)

RCT_EXPORT_VIEW_PROPERTY(onDrawItemComplete, RCTBubblingEventBlock)

RCT_EXPORT_VIEW_PROPERTY(onDrawPointComplete, RCTBubblingEventBlock)

RCT_EXPORT_VIEW_PROPERTY(optionList, NSString)

RCT_EXTERN_METHOD(updateLastCandlestick:(nonnull NSNumber *)node
                  candlestick:(nonnull NSDictionary *)candlestick)

RCT_EXTERN_METHOD(addCandlesticksAtTheEnd:(nonnull NSNumber *)node
                  candlesticks:(nonnull NSArray *)candlesticks)

RCT_EXTERN_METHOD(addCandlesticksAtTheStart:(nonnull NSNumber *)node
                  candlesticks:(nonnull NSArray *)candlesticks)

RCT_EXTERN_METHOD(addOrderLine:(nonnull NSNumber *)node
                  orderLine:(nonnull NSDictionary *)orderLine)

RCT_EXTERN_METHOD(removeOrderLine:(nonnull NSNumber *)node
                  orderLineId:(nonnull NSString *)orderLineId)

RCT_EXTERN_METHOD(updateOrderLine:(nonnull NSNumber *)node
                  orderLine:(nonnull NSDictionary *)orderLine)

RCT_EXTERN_METHOD(getOrderLines:(nonnull NSNumber *)node)

RCT_EXTERN_METHOD(addBuySellMark:(nonnull NSNumber *)node
                  buySellMark:(nonnull NSDictionary *)buySellMark)

RCT_EXTERN_METHOD(removeBuySellMark:(nonnull NSNumber *)node
                  buySellMarkId:(nonnull NSString *)buySellMarkId)

RCT_EXTERN_METHOD(updateBuySellMark:(nonnull NSNumber *)node
                  buySellMark:(nonnull NSDictionary *)buySellMark)

RCT_EXTERN_METHOD(getBuySellMarks:(nonnull NSNumber *)node)

@end

File: ios/Classes/RNKLineView.swift
//
//  HTKLineViewManager.swift
//  Base64
//
//  Created by hublot on 2020/4/3.
//

import UIKit

@objc(RNKLineView)
@objcMembers
class RNKLineView: RCTViewManager {

    static let queue = DispatchQueue.init(label: "com.hublot.klinedata")

    override func view() -> UIView! {
        return HTKLineContainerView()
    }

    override class func requiresMainQueueSetup() -> Bool {
        return true
    }

    @objc func updateLastCandlestick(_ node: NSNumber, candlestick: NSDictionary) {
        print("RNKLineView: updateLastCandlestick called for node \(node) with data: \(candlestick)")
        DispatchQueue.main.async {
            guard let view = self.bridge?.uiManager.view(forReactTag: node) as? HTKLineContainerView else {
                print("RNKLineView: Could not find HTKLineContainerView for node \(node)")
                return
            }
            print("RNKLineView: Calling view.updateLastCandlestick")
            view.updateLastCandlestick(candlestick)
        }
    }

    @objc func addCandlesticksAtTheEnd(_ node: NSNumber, candlesticks: NSArray) {
        print("RNKLineView: addCandlesticksAtTheEnd called for node \(node) with \(candlesticks.count) candlesticks")
        DispatchQueue.main.async {
            guard let view = self.bridge?.uiManager.view(forReactTag: node) as? HTKLineContainerView else {
                print("RNKLineView: Could not find HTKLineContainerView for node \(node)")
                return
            }
            print("RNKLineView: Calling view.addCandlesticksAtTheEnd")
            view.addCandlesticksAtTheEnd(candlesticks)
        }
    }

    @objc func addCandlesticksAtTheStart(_ node: NSNumber, candlesticks: NSArray) {
        print("RNKLineView: addCandlesticksAtTheStart called for node \(node) with \(candlesticks.count) candlesticks")
        DispatchQueue.main.async {
            guard let view = self.bridge?.uiManager.view(forReactTag: node) as? HTKLineContainerView else {
                print("RNKLineView: Could not find HTKLineContainerView for node \(node)")
                return
            }
            print("RNKLineView: Calling view.addCandlesticksAtTheStart")
            view.addCandlesticksAtTheStart(candlesticks)
        }
    }

    @objc func addOrderLine(_ node: NSNumber, orderLine: NSDictionary) {
        print("RNKLineView: addOrderLine called for node \(node) with data: \(orderLine)")
        DispatchQueue.main.async {
            guard let view = self.bridge?.uiManager.view(forReactTag: node) as? HTKLineContainerView else {
                print("RNKLineView: Could not find HTKLineContainerView for node \(node)")
                return
            }
            print("RNKLineView: Calling view.addOrderLine")
            view.addOrderLine(orderLine)
        }
    }

    @objc func removeOrderLine(_ node: NSNumber, orderLineId: NSString) {
        print("RNKLineView: removeOrderLine called for node \(node) with id: \(orderLineId)")
        DispatchQueue.main.async {
            guard let view = self.bridge?.uiManager.view(forReactTag: node) as? HTKLineContainerView else {
                print("RNKLineView: Could not find HTKLineContainerView for node \(node)")
                return
            }
            print("RNKLineView: Calling view.removeOrderLine")
            view.removeOrderLine(orderLineId as String)
        }
    }

    @objc func updateOrderLine(_ node: NSNumber, orderLine: NSDictionary) {
        print("RNKLineView: updateOrderLine called for node \(node) with data: \(orderLine)")
        DispatchQueue.main.async {
            guard let view = self.bridge?.uiManager.view(forReactTag: node) as? HTKLineContainerView else {
                print("RNKLineView: Could not find HTKLineContainerView for node \(node)")
                return
            }
            print("RNKLineView: Calling view.updateOrderLine")
            view.updateOrderLine(orderLine)
        }
    }

    @objc func getOrderLines(_ node: NSNumber) -> NSArray {
        print("RNKLineView: getOrderLines called for node \(node)")
        guard let view = self.bridge?.uiManager.view(forReactTag: node) as? HTKLineContainerView else {
            print("RNKLineView: Could not find HTKLineContainerView for node \(node)")
            return NSArray()
        }
        print("RNKLineView: Calling view.getOrderLines")
        return view.getOrderLines()
    }

    @objc func addBuySellMark(_ node: NSNumber, buySellMark: NSDictionary) {
        DispatchQueue.main.async {
            guard let view = self.bridge?.uiManager.view(forReactTag: node) as? HTKLineContainerView else {
                print("RNKLineView: Could not find HTKLineContainerView for node \(node)")
                return
            }
            view.addBuySellMark(buySellMark)
        }
    }

    @objc func removeBuySellMark(_ node: NSNumber, buySellMarkId: NSString) {
        DispatchQueue.main.async {
            guard let view = self.bridge?.uiManager.view(forReactTag: node) as? HTKLineContainerView else {
                print("RNKLineView: Could not find HTKLineContainerView for node \(node)")
                return
            }
            view.removeBuySellMark(buySellMarkId as String)
        }
    }

    @objc func updateBuySellMark(_ node: NSNumber, buySellMark: NSDictionary) {
        DispatchQueue.main.async {
            guard let view = self.bridge?.uiManager.view(forReactTag: node) as? HTKLineContainerView else {
                print("RNKLineView: Could not find HTKLineContainerView for node \(node)")
                return
            }
            view.updateBuySellMark(buySellMark)
        }
    }

    @objc func getBuySellMarks(_ node: NSNumber) -> NSArray {
        guard let view = self.bridge?.uiManager.view(forReactTag: node) as? HTKLineContainerView else {
            print("RNKLineView: Could not find HTKLineContainerView for node \(node)")
            return NSArray()
        }
        return view.getBuySellMarks()
    }

}
